cajita/src/Cajita_ParticleDynamicPartitioner.hpp

```diff
@@ -0,0 +1,112 @@
+/****************************************************************************
+ * Copyright (c) 2018-2022 by the Cabana authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the Cabana library. Cabana is distributed under a   *
+ * BSD 3-clause license. For the licensing terms see the LICENSE file in    *
+ * the top-level directory.                                                 *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+/*!
+  \file Cajita_ParticleDynamicPartitioner.hpp
+  \brief Multi-node particle based dynamic grid partitioner
+*/
+#ifndef CAJITA_PARTICLEDYNAMICPARTITIONER_HPP
+#define CAJITA_PARTICLEDYNAMICPARTITIONER_HPP
+
+#include <Cajita_DynamicPartitioner.hpp>
+#include <Cajita_SparseIndexSpace.hpp>
+#include <Kokkos_Core.hpp>
+
+#include <array>
+#include <vector>
+
+#include <mpi.h>
+
+namespace Cajita
+{
+
+template <class ParticlePosViewType, typename ArrayType, typename CellUnit,
+          unsigned long long CellPerTileDim, int num_space_dim, typename Device>
+class ParticleWorkloadSetter : public WorkloadSetter<Device>
+{
+    using memory_space = typename Device::memory_space;
+    using execution_space = typename Device::execution_space;
+
+    static constexpr unsigned long long cell_bits_per_tile_dim =
+        bitCount( CellPerTileDim );
+
+    const ParticlePosViewType& view;
+    int particle_num;
+    const ArrayType& global_lower_corner;
+    const CellUnit dx;
+    MPI_Comm comm;
+
+  public:
+    ParticleWorkloadSetter( const ParticlePosViewType& view, int particle_num,
+                            const ArrayType& global_lower_corner,
+                            const CellUnit dx, MPI_Comm comm )
+        : view( view )
+        , particle_num( particle_num )
+        , global_lower_corner( global_lower_corner )
+        , dx( dx )
+        , comm( comm )
+    {
+    }
+
+    void run( Kokkos::View<int***, memory_space>& workload ) override
+    {
+        Kokkos::Array<CellUnit, num_space_dim> lower_corner;
+        for ( std::size_t d = 0; d < num_space_dim; ++d )
+        {
+            lower_corner[d] = global_lower_corner[d];
+        }
+
+        Kokkos::parallel_for(
+            "compute_local_workload_parpos",
+            Kokkos::RangePolicy<execution_space>( 0, particle_num ),
+            KOKKOS_LAMBDA( const int i ) {
+                int ti = static_cast<int>(
+                             ( view( i, 0 ) - lower_corner[0] ) / dx - 0.5 ) >>
```

Comment:
We'll either need to make this a class functor or create a local function copies of the member variables - capturing members in a lambda doesn't work in many cases. Either is fine with me